src/core/Instrument.cpp

```diff
@@ -79,7 +81,52 @@ bool Instrument::isFromTrack( const Track * _track ) const
 }
 
 
-
+void Instrument::applyFadeIn(  sampleFrame * buf, const NotePlayHandle * _n )
+{
+    // apply only if it's the start of the note
+    if( _n->totalFramesPlayed() == 0 ){
+        const fpp_t frames = _n->framesLeftForCurrentPeriod();
+
+        // zero point crossing counts of all channels
+        int zero_crossings[DEFAULT_CHANNELS] = {0};
+        // maximum zero point crossing of all channels
+        int max_zc = 0;
+        
+        // determine the zero point crossing counts
+        for( fpp_t f = 0; f < frames; ++f )
+        {
+            for( ch_cnt_t ch=0; ch < DEFAULT_CHANNELS; ++ch )
+            {
+                if( ( buf[f-1][ch] < 0.0 && buf[f][ch] > 0.0 ) ||
+                    ( buf[f-1][ch] > 0.0 && buf[f][ch] < 0.0 ) )
+                {
+                    ++zero_crossings[ch];
+                }
+            }
+        }
+
+        // find the biggest one
+        for( ch_cnt_t ch = 0; ch < DEFAULT_CHANNELS; ++ch )
+        {
+            if( zero_crossings[ch] > max_zc )
+            {
+                max_zc = zero_crossings[ch];
+            }
+        }
```

Comment:
This loop is pulled out of the previous, which to me says this `if` check can be combined into the previous for loop to prevent looping through things multiple times.

Reply:
I considered that but I was unsure whether it will be more efficient or not. Because if we combine these loops, it will check if it's the new maximum for every zero point crossing. Performance difference is probably marginal, but perhaps it can make a meaningful difference if there are too many instruments.

I will merge these two loops anyway in the next commit to improve readability. If someone proves that the current method is faster, I can revert these changes.